OptimizelySDK.Tests/DecisionServiceTest.cs

```diff
@@ -574,9 +572,156 @@ public void TestGetVariationForFeatureWhenTheUserIsNotBucketedIntoFeatureExperim
+        [Test]
+        public void TestGetVariationForFeatureRolloutWhenUserIsBucketedInTheTargetingRule()
+        {
+            var featureFlag = ProjectConfig.GetFeatureFlagFromKey("boolean_single_variable_feature");
+            var rolloutId = featureFlag.RolloutId;
+            var rollout = ProjectConfig.GetRolloutFromId(rolloutId);
+            var experiment = rollout.Experiments[0];
+            var expectedVariation = experiment.Variations[0];
+
+            var userAttributes = new UserAttributes {
+                { "browser_type", "chrome" }
+            };
+
+            BucketerMock.Setup(bm => bm.Bucket(It.IsAny<ProjectConfig>(), It.IsAny<Experiment>(), It.IsAny<string>(), It.IsAny<string>())).Returns(expectedVariation);
+            var decisionService = new DecisionService(BucketerMock.Object, ErrorHandlerMock.Object, ProjectConfig, null, LoggerMock.Object);
+
+            var actualVariation = decisionService.GetVariationForFeatureRollout(featureFlag, "user_1", userAttributes);
+            Assert.AreEqual(expectedVariation, actualVariation);
+
+            LoggerMock.Verify(l => l.Log(LogLevel.DEBUG, $"Attempting to bucket user \"user_1\" into rollout rule \"{experiment.Key}\"."));
+        }
+
+        // Should return the variation the user is bucketed into when the user is bucketed into the "Everyone Else" rule
+        // and the user is not bucketed into the targeting rule
+        [Test]
+        public void TestGetVariationForFeatureRolloutWhenUserIsNotBucketedInTheTargetingRuleButBucketedToEveryoneElseRule()
+        {
+            var featureFlag = ProjectConfig.GetFeatureFlagFromKey("boolean_single_variable_feature");
+            var rolloutId = featureFlag.RolloutId;
+            var rollout = ProjectConfig.GetRolloutFromId(rolloutId);
+            var experiment = rollout.Experiments[0];
+            var everyoneElseRule = rollout.Experiments[rollout.Experiments.Count - 1];
+            var expectedVariation = everyoneElseRule.Variations[0];
+
+            var userAttributes = new UserAttributes {
+                { "browser_type", "chrome" }
+            };
+
+            BucketerMock.Setup(bm => bm.Bucket(It.IsAny<ProjectConfig>(), experiment, It.IsAny<string>(), It.IsAny<string>())).Returns<Variation>(null);
+            BucketerMock.Setup(bm => bm.Bucket(It.IsAny<ProjectConfig>(), everyoneElseRule, It.IsAny<string>(), It.IsAny<string>())).Returns(expectedVariation);
+            var decisionService = new DecisionService(BucketerMock.Object, ErrorHandlerMock.Object, ProjectConfig, null, LoggerMock.Object);
+
+            var actualVariation = decisionService.GetVariationForFeatureRollout(featureFlag, "user_1", userAttributes);
+            Assert.AreEqual(expectedVariation, actualVariation);
+
+            LoggerMock.Verify(l => l.Log(LogLevel.DEBUG, $"Attempting to bucket user \"user_1\" into rollout rule \"{experiment.Key}\"."));
```

Comment:
what is this $?

Reply:
Sitting in US and asking about $ from Paskistani people is a bad idea ! 😛 

$ adds capability to use variables and even functions inside raw strings. It eliminates the need of string.format method to some extent.

$"Attempting to bucket user \"user_1\" into rollout rule \"{experiment.Key}\"."
In the example above, I called {experiment.Key}  inside the string. Without $, the line would become something like this:
string.Format("Attempting to bucket user \"user_1\" into rollout rule \"{1}\".", experiment.Key);